core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -66,169 +64,164 @@ class MetadataCache(brokerId: Int) extends Logging {
+              new MetadataResponse.PartitionMetadata(Errors.REPLICA_NOT_AVAILABLE, partitionId.toInt, leader,
                 replicaInfo.asJava, isrInfo.asJava, offlineReplicaInfo.asJava)
             } else {
-              new MetadataResponse.PartitionMetadata(Errors.NONE, partitionId, leader, replicaInfo.asJava,
+              new MetadataResponse.PartitionMetadata(Errors.NONE, partitionId.toInt, leader, replicaInfo.asJava,
                 isrInfo.asJava, offlineReplicaInfo.asJava)
             }
         }
       }
     }
   }
 
-  private def getAliveEndpoint(brokerId: Int, listenerName: ListenerName): Option[Node] =
-    inReadLock(partitionMetadataLock) {
-      // Returns None if broker is not alive or if the broker does not have a listener named `listenerName`.
-      // Since listeners can be added dynamically, a broker with a missing listener could be a transient error.
-      aliveNodes.get(brokerId).flatMap(_.get(listenerName))
-    }
+  private def getAliveEndpoint(snapshot: MetadataSnapshot, brokerId: Int, listenerName: ListenerName): Option[Node] =
+    // Returns None if broker is not alive or if the broker does not have a listener named `listenerName`.
+    // Since listeners can be added dynamically, a broker with a missing listener could be a transient error.
+    snapshot.aliveNodes.get(brokerId).flatMap(_.get(listenerName))
 
   // errorUnavailableEndpoints exists to support v0 MetadataResponses
   def getTopicMetadata(topics: Set[String], listenerName: ListenerName, errorUnavailableEndpoints: Boolean = false,
                        errorUnavailableListeners: Boolean = false): Seq[MetadataResponse.TopicMetadata] = {
-    inReadLock(partitionMetadataLock) {
-      topics.toSeq.flatMap { topic =>
-        getPartitionMetadata(topic, listenerName, errorUnavailableEndpoints, errorUnavailableListeners).map { partitionMetadata =>
-          new MetadataResponse.TopicMetadata(Errors.NONE, topic, Topic.isInternal(topic), partitionMetadata.toBuffer.asJava)
-        }
+    val snapshot = metadataSnapshot
+    topics.toSeq.flatMap { topic =>
+      getPartitionMetadata(snapshot, topic, listenerName, errorUnavailableEndpoints, errorUnavailableListeners).map { partitionMetadata =>
+        new MetadataResponse.TopicMetadata(Errors.NONE, topic, Topic.isInternal(topic), partitionMetadata.toBuffer.asJava)
       }
     }
   }
 
   def getAllTopics(): Set[String] = {
-    inReadLock(partitionMetadataLock) {
-      cache.keySet.toSet
-    }
+    getAllTopics(metadataSnapshot)
   }
 
-  def getAllPartitions(): Map[TopicPartition, UpdateMetadataRequest.PartitionState] = {
-    inReadLock(partitionMetadataLock) {
-      cache.flatMap { case (topic, partitionStates) =>
-        partitionStates.map { case (partition, state ) => (new TopicPartition(topic, partition), state) }
-      }.toMap
-    }
+  private def getAllTopics(snapshot: MetadataSnapshot): Set[String] = {
+    snapshot.partitionStates.keySet.toSet
```

Comment:
nits: could we remove the `toSet` here?

Reply:
Yeah, we are doing a copy to an immutable `Set` here. Since we never mutate the cached data, we can probably remove that and make this operation very cheap instead of O(n).

Reply:
good catch